src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -69,12 +61,14 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
     def genStat(tree: Tree) {
       lineNumber(tree)
       tree match {
-        case Assign(lhs @ Select(_, _), rhs) =>
+        case Assign(lhs @ Select(qual, _), rhs) =>
           val isStatic = lhs.symbol.isStaticMember
           if (!isStatic) { genLoadQualifier(lhs) }
           genLoad(rhs, symInfoTK(lhs.symbol))
           lineNumber(tree)
-          fieldStore(lhs.symbol)
+          // receiverClass is used in the bytecode to access the field. using sym.owner may lead to IllegalAccessError, SI-4283
+          val receiverClass = qual.tpe.typeSymbol
```

Comment:
Scala doesn't allow static references via an instance or a subclass qualifier. I just noticed that this makes it impossible to express some static field/method selections.

```scala
class D {
  def test(b: Test.B) {
    b.i = b.i + 1;

    b.j = b.j + 1; // j not a member
    Test.B.j = Test.B.j + 1; // j not a member
    Test.A.j = Test.A.j + 1; // A is not accessible

    b.m();
    b.n(); // n is not a member
    Test.B.n(); // n is not a member
    Test.A.n(); // A is not accessible
  }
}
```